### PR TITLE
Promote OkHttp's Pipe into Okio.

### PR DESCRIPTION
**Pull request by swankjesse**
Also make Timeout.waitUntilNotified() a public API.

### CODE REVIEW

File: okio/src/main/java/okio/Pipe.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  final Source source = new PipeSource();
+
+  public Pipe(long maxBufferSize) {
+    if (maxBufferSize < 1L) {
+      throw new IllegalArgumentException("maxBufferSize < 1: " + maxBufferSize);
+    }
+    this.maxBufferSize = maxBufferSize;
+  }
+
+  public Source source() {
```
Comment (swankjesse) [inline okio/src/main/java/okio/Pipe.java]: New accessors so that `sink`, `source` fields don’t need to be public.

Comment (swankjesse) [inline okio/src/main/java/okio/Pipe.java]: . . . thoughts? The public members had a very simple struct feel to it.

File: okio/src/main/java/okio/Timeout.java [hunk 2]
```diff
@@ -149,4 +150,48 @@ public void throwIfReached() throws IOException {
       throw new InterruptedIOException("deadline reached");
     }
   }
+
+  /**
+   * Waits on {@code monitor} until it is notified. Throws {@link SocketTimeoutException} if either
+   * the thread is interrupted or if this timeout elapses before {@code monitor} is notified. The
+   * caller must be synchronized on {@code monitor}.
+   */
+  public final void waitUntilNotified(Object monitor) throws IOException {
```
Comment (swankjesse) [inline okio/src/main/java/okio/Timeout.java]: (In OkHttp this is a static method that takes a `Timeout` as a parameter.)

File: okio/src/main/java/okio/Pipe.java (hunk 1 above)
Comment (JakeWharton) [inline okio/src/main/java/okio/Pipe.java]: Doesn't bother me. Feels more appropriate here anyway.

File: okio/src/main/java/okio/Timeout.java [hunk 3]
```diff
@@ -149,4 +150,79 @@ public void throwIfReached() throws IOException {
...
+        monitor.wait(waitMillis, (int) (waitNanos - waitMillis * 1000000L));
+        elapsedNanos = System.nanoTime() - start;
+      }
+
+      // Throw if the timeout elapsed before the monitor was notified.
+      if (elapsedNanos >= waitNanos) {
+        throw new SocketTimeoutException("timeout");
+      }
+    } catch (InterruptedException e) {
+      throw new SocketTimeoutException("interrupted");
```
Comment (JakeWharton) [inline okio/src/main/java/okio/Timeout.java]: The use of `SocketTimeoutException` in here is unfortunate. Our desired use is leaking. How about letting `InterruptedException` propagate and returning a `boolean` as to whether timeout occurred?

File: okio/src/main/java/okio/Pipe.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+ * been exhausted. At that point reads will return -1, indicating the end of the stream. But if the
+ * source is closed first, writes to the sink will immediately fail with an {@link IOException}.
+ */
+public final class Pipe {
+  final long maxBufferSize;
+  final Buffer buffer = new Buffer();
+  boolean sinkClosed;
+  boolean sourceClosed;
+  final Sink sink = new PipeSink();
+  final Source source = new PipeSource();
```
Comment (JakeWharton) [inline okio/src/main/java/okio/Pipe.java]: Make these two `private`. Force tests to use more of the public API.

**Comment by JakeWharton** [comment]
lgtm. i only scanned the tests since i reviewed them in detail going into OkHttp and I assume they're the same.

File: okio/src/main/java/okio/Timeout.java (hunk 3 above)
Comment (swankjesse) [inline okio/src/main/java/okio/Timeout.java]: Yeah, this isn’t very good. I was cheating and treating `SocketTimeoutException ` as a `TimeoutException` which is a small lie. Still, I don’t like returning a boolean either here. This thing is trying to package up a timeout abstraction and throwing when it elapses is part of that.

What about keeping the SocketTimeoutException above, and trading this for an InterruptedIOException?

File: okio/src/main/java/okio/Pipe.java (hunk 4 above)
Comment (swankjesse) [inline okio/src/main/java/okio/Pipe.java]: fixed!

File: okio/src/main/java/okio/Timeout.java (hunk 3 above)
Comment (JakeWharton) [inline okio/src/main/java/okio/Timeout.java]: `throwIfReached` uses `InterruptedIOException` for both cases. Should this then? 

As to your suggestions specifically, I'm not sure if two checked exceptions is strictly worse than one whose name isn't quite perfect.

**Comment by swankjesse** [comment]
https://github.com/square/okio/issues/218

File: okio/src/main/java/okio/Timeout.java (hunk 3 above)
Comment (swankjesse) [inline okio/src/main/java/okio/Timeout.java]: InterruptedIOException works for me. Note that SocketTimeoutException is a subclass of that!

File: okio/src/main/java/okio/Timeout.java [hunk 5]
```diff
@@ -149,4 +150,79 @@ public void throwIfReached() throws IOException {
       throw new InterruptedIOException("deadline reached");
     }
   }
+
+  /**
+   * Waits on {@code monitor} until it is notified. Throws {@link SocketTimeoutException} if either
```
Comment (JakeWharton) [inline okio/src/main/java/okio/Timeout.java]: Don't forget to change this javadoc

**Pull request merged**